README
General Information
===================

FUSE (Filesystem in Userspace) is a simple interface for userspace
programs to export a virtual filesystem to the Linux kernel.  FUSE
also aims to provide a secure method for non privileged users to
create and mount their own filesystem implementations.

You can download the source code releases from

  http://sourceforge.net/projects/fuse

or alternatively you can use CVS to get the very latest development
version:

  cvs -d :pserver:anonymous@fuse.cvs.sourceforge.net:/cvsroot/fuse co fuse


Dependencies
============

Linux kernel version 2.6.X where X >= 9.

Alternatively a kernel module from FUSE release 2.5.* can be used with
this release, which supports kernels >= 2.4.21.

Installation
============

sudo apt-get install automake libtool
./makeconf.sh
./configure --libdir="/lib/x86_64-linux-gnu"
make
sudo make install

Assuming you have cloned and installed tm-fuse-4.x.x,
modprobe tm-fuse

You may also need to add '/usr/local/lib' to '/etc/ld.so.conf' and/or
run ldconfig.

You'll also need a fuse kernel module, Linux kernels 2.6.14 or later
contain FUSE support.

For more details see the file 'INSTALL'

How To Use
==========

FUSE is made up of three main parts:

 - A kernel filesystem module

 - A userspace library

 - A mount/unmount program


Here's how to create your very own virtual filesystem in five easy
steps (after installing FUSE):

  1) Edit the file example/fusexmp.c to do whatever you want...

  2) Build the fusexmp program

  3) run 'example/fusexmp /mnt/fuse -d'

  4) ls -al /mnt/fuse

  5) Be glad

If it doesn't work out, please ask!  Also see the file 'include/fuse.h' for
detailed documentation of the library interface.

Security
========

If you run 'make install', the fusermount program is installed
set-user-id to root.  This is done to allow normal users to mount
their own filesystem implementations.

There must however be some limitations, in order to prevent Bad User from
doing nasty things.  Currently those limitations are:

  - The user can only mount on a mountpoint, for which it has write
    permission

  - The mountpoint is not a sticky directory which isn't owned by the
    user (like /tmp usually is)

  - No other user (including root) can access the contents of the mounted
    filesystem.

Configuration
=============

Some options regarding mount policy can be set in the file
'/etc/fuse.conf'

Currently these options are:

mount_max = NNN

  Set the maximum number of FUSE mounts allowed to non-root users.
  The default is 1000.

user_allow_other

  Allow non-root users to specify the 'allow_other' or 'allow_root'
  mount options.


Mount options
=============

Most of the generic mount options described in 'man mount' are
supported (ro, rw, suid, nosuid, dev, nodev, exec, noexec, atime,
noatime, sync async, dirsync).  Filesystems are mounted with
'-onodev,nosuid' by default, which can only be overridden by a
privileged user.

These are FUSE specific mount options that can be specified for all
filesystems:

default_permissions

  By default FUSE doesn't check file access permissions, the
  filesystem is free to implement it's access policy or leave it to
  the underlying file access mechanism (e.g. in case of network
  filesystems).  This option enables permission checking, restricting
  access based on file mode.  This is option is usually useful
  together with the 'allow_other' mount option.

allow_other

  This option overrides the security measure restricting file access
  to the user mounting the filesystem.  So all users (including root)
  can access the files.  This option is by default only allowed to
  root, but this restriction can be removed with a configuration
  option described in the previous section.

allow_root

  This option is similar to 'allow_other' but file access is limited
  to the user mounting the filesystem and root.  This option and
  'allow_other' are mutually exclusive.

kernel_cache

  This option disables flushing the cache of the file contents on
  every open().  This should only be enabled on filesystems, where the
  file data is never changed externally (not through the mounted FUSE
  filesystem).  Thus it is not suitable for network filesystems and
  other "intermediate" filesystems.

  NOTE: if this option is not specified (and neither 'direct_io') data
  is still cached after the open(), so a read() system call will not
  always initiate a read operation.

auto_cache

  This option enables automatic flushing of the data cache on open().
  The cache will only be flushed if the modification time or the size
  of the file has changed.

large_read

  Issue large read requests.  This can improve performance for some
  filesystems, but can also degrade performance.  This option is only
  useful on 2.4.X kernels, as on 2.6 kernels requests size is
  automatically determined for optimum performance.

direct_io

  This option disables the use of page cache (file content cache) in
  the kernel for this filesystem.  This has several affects:

     - Each read() or write() system call will initiate one or more
       read or write operations, data will not be cached in the
       kernel.

     - The return value of the read() and write() system calls will
       correspond to the return values of the read and write
       operations.  This is useful for example if the file size is not
       known in advance (before reading it).

max_read=N

  With this option the maximum size of read operations can be set.
  The default is infinite.  Note that the size of read requests is
  limited anyway to 32 pages (which is 128kbyte on i386).

max_readahead=N

  Set the maximum number of bytes to read-ahead.  The default is
  determined by the kernel.  On linux-2.6.22 or earlier it's 131072
  (128kbytes)

max_write=N

  Set the maximum number of bytes in a single write operation.  The
  default is 128kbytes.  Note, that due to various limitations, the
  size of write requests can be much smaller (4kbytes).  This
  limitation will be removed in the future.

async_read

  Perform reads asynchronously. This is the default

sync_read

  Perform all reads (even read-ahead) synchronously.

hard_remove

  The default behavior is that if an open file is deleted, the file is
  renamed to a hidden file (.fuse_hiddenXXX), and only removed when
  the file is finally released.  This relieves the filesystem
  implementation of having to deal with this problem.  This option
  disables the hiding behavior, and files are removed immediately in
  an unlink operation (or in a rename operation which overwrites an
  existing file).

  It is recommended that you not use the hard_remove option. When
  hard_remove is set, the following libc functions fail on unlinked
  files (returning errno of ENOENT):
     - read()
     - write()
     - fsync()
     - close()
     - f*xattr()
     - ftruncate()
     - fstat()
     - fchmod()
     - fchown()

debug

  Turns on debug information printing by the library.

fsname=NAME

  Sets the filesystem source (first field in /etc/mtab).  The default
  is the program name.

subtype=TYPE

  Sets the filesystem type (third field in /etc/mtab).  The default is
  the program name.

  If the kernel suppports it, /etc/mtab and /proc/mounts will show the
  filesystem type as "fuse.TYPE"

  If the kernel doesn't support subtypes, the source filed will be
  "TYPE#NAME", or if fsname option is not specified, just "TYPE".

use_ino

  Honor the 'st_ino' field in getattr() and fill_dir().  This value is
  used to fill in the 'st_ino' field in the stat()/lstat()/fstat()
  functions and the 'd_ino' field in the readdir() function.  The
  filesystem does not have to guarantee uniqueness, however some
  applications rely on this value being unique for the whole
  filesystem.

readdir_ino

  If 'use_ino' option is not given, still try to fill in the 'd_ino'
  field in readdir().  If the name was previously looked up, and is
  still in the cache, the inode number found there will be used.
  Otherwise it will be set to '-1'.  If 'use_ino' option is given,
  this option is ignored.

nonempty

  Allows mounts over a non-empty file or directory.  By default these
  mounts are rejected (from version 2.3.1) to prevent accidental
  covering up of data, which could for example prevent automatic
  backup.

umask=M

  Override the permission bits in 'st_mode' set by the filesystem.
  The resulting permission bits are the ones missing from the given
  umask value.  The value is given in octal representation.

uid=N

  Override the 'st_uid' field set by the filesystem.

gid=N

  Override the 'st_gid' field set by the filesystem.

blkdev

  Mount a filesystem backed by a block device.  This is a privileged
  option.  The device must be specified with the 'fsname=NAME' option.

entry_timeout=T

  The timeout in seconds for which name lookups will be cached. The
  default is 1.0 second.  For all the timeout options, it is possible
  to give fractions of a second as well (e.g. "-oentry_timeout=2.8")

negative_timeout=T

  The timeout in seconds for which a negative lookup will be cached.
  This means, that if file did not exist (lookup retuned ENOENT), the
  lookup will only be redone after the timeout, and the file/directory
  will be assumed to not exist until then.  The default is 0.0 second,
  meaning that caching negative lookups are disabled.

attr_timeout=T

  The timeout in seconds for which file/directory attributes are
  cached.  The default is 1.0 second.

ac_attr_timeout=T

  The timeout in seconds for which file attributes are cached for the
  purpose of checking if "auto_cache" should flush the file data on
  open.   The default is the value of 'attr_timeout'

intr

  Allow requests to be interrupted.  Turning on this option may result
  in unexpected behavior, if the filesystem does not support request
  interruption.

intr_signal=NUM

  Specify which signal number to send to the filesystem when a request
  is interrupted.  The default is 10 (USR1).

modules=M1[:M2...]

  Add modules to the filesystem stack.  Modules are pushed in the
  order they are specified, with the original filesystem being on the
  bottom of the stack.


Modules distributed with fuse
-----------------------------

iconv
`````
Perform file name character set conversion.  Options are:

from_code=CHARSET

  Character set to convert from (see iconv -l for a list of possible
  values).  Default is UTF-8.

to_code=CHARSET

  Character set to convert to.  Default is determined by the current
  locale.


subdir
``````
Prepend a given directory to each path. Options are:

subdir=DIR

  Directory to prepend to all paths.  This option is mandatory.

rellinks

  Transform absolute symlinks into relative

norellinks

  Do not transform absolute symlinks into relative.  This is the default.


Reporting bugs
==============

Please send bug reports to the <fuse-devel@lists.sourceforge.net>
mailing list.

The list is open, you need not be subscribed to post.


debian/changelog
tm-libfuse (1.1-1) unstable; urgency=low

  * Repackage for unstable

 -- Keith Packard <packard@hpe.com>  Tue, 15 Nov 2016 14:57:10 -0800


lib/mount.c
/*
  TMFS: Filesystem in Userspace
  Copyright (C) 2001-2007  Miklos Szeredi <miklos@szeredi.hu>

  This program can be distributed under the terms of the GNU LGPLv2.
  See the file COPYING.LIB.
*/

#include "config.h"
#include "tmfs_i.h"
#include "tmfs_misc.h"
#include "tmfs_opt.h"
#include "tmfs_common_compat.h"
#include "mount_util.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stddef.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/poll.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/wait.h>
#include <sys/mount.h>

#ifdef __NetBSD__
#include <pertmfs.h>

#define MS_RDONLY 	MNT_RDONLY
#define MS_NOSUID 	MNT_NOSUID
#define MS_NODEV 	MNT_NODEV
#define MS_NOEXEC 	MNT_NOEXEC
#define MS_SYNCHRONOUS 	MNT_SYNCHRONOUS
#define MS_NOATIME 	MNT_NOATIME

#define umount2(mnt, flags) unmount(mnt, (flags == 2) ? MNT_FORCE : 0)
#endif

#define TMFSRMOUNT_PROG		"tmfsrmount"
#define TMFS_COMMFD_ENV		"_TMFS_COMMFD"

#ifndef HAVE_FORK
#define fork() vfork()
#endif

#ifndef MS_DIRSYNC
#define MS_DIRSYNC 128
#endif

enum {
	KEY_KERN_FLAG,
	KEY_KERN_OPT,
	KEY_TMFSRMOUNT_OPT,
	KEY_SUBTYPE_OPT,
	KEY_MTAB_OPT,
	KEY_ALLOW_ROOT,
	KEY_RO,
	KEY_HELP,
	KEY_VERSION,
};

struct mount_opts {
	int allow_other;
	int allow_root;
	int ishelp;
	int flags;
	int nonempty;
	int auto_unmount;
	int blkdev;
	char *fsname;
	char *subtype;
	char *subtype_opt;
	char *mtab_opts;
	char *tmfsrmount_opts;
	char *kernel_opts;
};

#define TMFS_MOUNT_OPT(t, p) { t, offsetof(struct mount_opts, p), 1 }

static const struct tmfs_opt tmfs_mount_opts[] = {
	TMFS_MOUNT_OPT("allow_other",		allow_other),
	TMFS_MOUNT_OPT("allow_root",		allow_root),
	TMFS_MOUNT_OPT("nonempty",		nonempty),
	TMFS_MOUNT_OPT("blkdev",		blkdev),
	TMFS_MOUNT_OPT("auto_unmount",		auto_unmount),
	TMFS_MOUNT_OPT("fsname=%s",		fsname),
	TMFS_MOUNT_OPT("subtype=%s",		subtype),
	TMFS_OPT_KEY("allow_other",		KEY_KERN_OPT),
	TMFS_OPT_KEY("allow_root",		KEY_ALLOW_ROOT),
	TMFS_OPT_KEY("nonempty",		KEY_TMFSRMOUNT_OPT),
	TMFS_OPT_KEY("auto_unmount",		KEY_TMFSRMOUNT_OPT),
	TMFS_OPT_KEY("blkdev",			KEY_TMFSRMOUNT_OPT),
	TMFS_OPT_KEY("fsname=",			KEY_TMFSRMOUNT_OPT),
	TMFS_OPT_KEY("subtype=",		KEY_SUBTYPE_OPT),
	TMFS_OPT_KEY("large_read",		KEY_KERN_OPT),
	TMFS_OPT_KEY("blksize=",		KEY_KERN_OPT),
	TMFS_OPT_KEY("default_permissions",	KEY_KERN_OPT),
	TMFS_OPT_KEY("max_read=",		KEY_KERN_OPT),
	TMFS_OPT_KEY("max_read=",		TMFS_OPT_KEY_KEEP),
	TMFS_OPT_KEY("user=",			KEY_MTAB_OPT),
	TMFS_OPT_KEY("-r",			KEY_RO),
	TMFS_OPT_KEY("ro",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("rw",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("suid",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("nosuid",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("dev",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("nodev",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("exec",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("noexec",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("async",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("sync",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("dirsync",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("atime",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("noatime",			KEY_KERN_FLAG),
	TMFS_OPT_KEY("-h",			KEY_HELP),
	TMFS_OPT_KEY("--help",			KEY_HELP),
	TMFS_OPT_KEY("-V",			KEY_VERSION),
	TMFS_OPT_KEY("--version",		KEY_VERSION),
	TMFS_OPT_END
};

static void mount_help(void)
{
	fprintf(stderr,
"    -o allow_other         allow access to other users\n"
"    -o allow_root          allow access to root\n"
"    -o auto_unmount        auto unmount on process termination\n"
"    -o nonempty            allow mounts over non-empty file/dir\n"
"    -o default_permissions enable permission checking by kernel\n"
"    -o fsname=NAME         set filesystem name\n"
"    -o subtype=NAME        set filesystem type\n"
"    -o large_read          issue large read requests (2.4 only)\n"
"    -o max_read=N          set maximum size of read requests\n"
"\n");
}

static void exec_tmfsrmount(const char *argv[])
{
	execv(TMFSRMOUNT_DIR "/" TMFSRMOUNT_PROG, (char **) argv);
	execvp(TMFSRMOUNT_PROG, (char **) argv);
}

static void mount_version(void)
{
	int pid = fork();
	if (!pid) {
		const char *argv[] = { TMFSRMOUNT_PROG, "--version", NULL };
		exec_tmfsrmount(argv);
		_exit(1);
	} else if (pid != -1)
		waitpid(pid, NULL, 0);
}

struct mount_flags {
	const char *opt;
	unsigned long flag;
	int on;
};

static const struct mount_flags mount_flags[] = {
	{"rw",	    MS_RDONLY,	    0},
	{"ro",	    MS_RDONLY,	    1},
	{"suid",    MS_NOSUID,	    0},
	{"nosuid",  MS_NOSUID,	    1},
	{"dev",	    MS_NODEV,	    0},
	{"nodev",   MS_NODEV,	    1},
	{"exec",    MS_NOEXEC,	    0},
	{"noexec",  MS_NOEXEC,	    1},
	{"async",   MS_SYNCHRONOUS, 0},
	{"sync",    MS_SYNCHRONOUS, 1},
	{"atime",   MS_NOATIME,	    0},
	{"noatime", MS_NOATIME,	    1},
#ifndef __NetBSD__
	{"dirsync", MS_DIRSYNC,	    1},
#endif
	{NULL,	    0,		    0}
};

static void set_mount_flag(const char *s, int *flags)
{
	int i;

	for (i = 0; mount_flags[i].opt != NULL; i++) {
		const char *opt = mount_flags[i].opt;
		if (strcmp(opt, s) == 0) {
			if (mount_flags[i].on)
				*flags |= mount_flags[i].flag;
			else
				*flags &= ~mount_flags[i].flag;
			return;
		}
	}
	fprintf(stderr, "tmfs: internal error, can't find mount flag\n");
	abort();
}

static int tmfs_mount_opt_proc(void *data, const char *arg, int key,
			       struct tmfs_args *outargs)
{
	struct mount_opts *mo = data;

	switch (key) {
	case KEY_ALLOW_ROOT:
		if (tmfs_opt_add_opt(&mo->kernel_opts, "allow_other") == -1 ||
		    tmfs_opt_add_arg(outargs, "-oallow_root") == -1)
			return -1;
		return 0;

	case KEY_RO:
		arg = "ro";
		/* fall through */
	case KEY_KERN_FLAG:
		set_mount_flag(arg, &mo->flags);
		return 0;

	case KEY_KERN_OPT:
		return tmfs_opt_add_opt(&mo->kernel_opts, arg);

	case KEY_TMFSRMOUNT_OPT:
		return tmfs_opt_add_opt_escaped(&mo->tmfsrmount_opts, arg);

	case KEY_SUBTYPE_OPT:
		return tmfs_opt_add_opt(&mo->subtype_opt, arg);

	case KEY_MTAB_OPT:
		return tmfs_opt_add_opt(&mo->mtab_opts, arg);

	case KEY_HELP:
		mount_help();
		mo->ishelp = 1;
		break;

	case KEY_VERSION:
		mount_version();
		mo->ishelp = 1;
		break;
	}
	return 1;
}

/* return value:
 * >= 0	 => fd
 * -1	 => error
 */
static int receive_fd(int fd)
{
	struct msghdr msg;
	struct iovec iov;
	char buf[1];
	int rv;
	size_t ccmsg[CMSG_SPACE(sizeof(int)) / sizeof(size_t)];
	struct cmsghdr *cmsg;

	iov.iov_base = buf;
	iov.iov_len = 1;

	memset(&msg, 0, sizeof(msg));
	msg.msg_name = 0;
	msg.msg_namelen = 0;
	msg.msg_iov = &iov;
	msg.msg_iovlen = 1;
	/* old BSD implementations should use msg_accrights instead of
	 * msg_control; the interface is different. */
	msg.msg_control = ccmsg;
	msg.msg_controllen = sizeof(ccmsg);

	while(((rv = recvmsg(fd, &msg, 0)) == -1) && errno == EINTR);
	if (rv == -1) {
		perror("recvmsg");
		return -1;
	}
	if(!rv) {
		/* EOF */
		return -1;
	}

	cmsg = CMSG_FIRSTHDR(&msg);
	if (!cmsg->cmsg_type == SCM_RIGHTS) {
		fprintf(stderr, "got control message of unknown type %d\n",
			cmsg->cmsg_type);
		return -1;
	}
	return *(int*)CMSG_DATA(cmsg);
}

void tmfs_kern_unmount(const char *mountpoint, int fd)
{
	int res;
	int pid;

	if (!mountpoint)
		return;

	if (fd != -1) {
		struct pollfd pfd;

		pfd.fd = fd;
		pfd.events = 0;
		res = poll(&pfd, 1, 0);

		/* Need to close file descriptor, otherwise synchronous umount
		   would recurse into filesystem, and deadlock.

		   Caller expects tmfs_kern_unmount to close the fd, so close it
		   anyway. */
		close(fd);

		/* If file poll returns POLLERR on the device file descriptor,
		   then the filesystem is already unmounted */
		if (res == 1 && (pfd.revents & POLLERR))
			return;
	}

	if (geteuid() == 0) {
		tmfs_mnt_umount("tmfs", mountpoint, mountpoint,  1);
		return;
	}

	res = umount2(mountpoint, 2);
	if (res == 0)
		return;

	pid = fork();
	if(pid == -1)
		return;

	if(pid == 0) {
		const char *argv[] = { TMFSRMOUNT_PROG, "-u", "-q", "-z",
				       "--", mountpoint, NULL };

		exec_tmfsrmount(argv);
		_exit(1);
	}
	waitpid(pid, NULL, 0);
}

void tmfs_unmount_compat22(const char *mountpoint)
{
	tmfs_kern_unmount(mountpoint, -1);
}

static int tmfs_mount_tmfsrmount(const char *mountpoint, struct mount_opts *mo,
		const char *opts, int quiet)
{
	int fds[2], pid;
	int res;
	int rv;

	if (!mountpoint) {
		fprintf(stderr, "tmfs: missing mountpoint parameter\n");
		return -1;
	}

	res = socketpair(PF_UNIX, SOCK_STREAM, 0, fds);
	if(res == -1) {
		perror("tmfs: socketpair() failed");
		return -1;
	}

	pid = fork();
	if(pid == -1) {
		perror("tmfs: fork() failed");
		close(fds[0]);
		close(fds[1]);
		return -1;
	}

	if(pid == 0) {
		char env[10];
		const char *argv[32];
		int a = 0;

		if (quiet) {
			int fd = open("/dev/null", O_RDONLY);
			if (fd != -1) {
				dup2(fd, 1);
				dup2(fd, 2);
			}
		}

		argv[a++] = TMFSRMOUNT_PROG;
		if (opts) {
			argv[a++] = "-o";
			argv[a++] = opts;
		}
		argv[a++] = "--";
		argv[a++] = mountpoint;
		argv[a++] = NULL;

		close(fds[1]);
		fcntl(fds[0], F_SETFD, 0);
		snprintf(env, sizeof(env), "%i", fds[0]);
		setenv(TMFS_COMMFD_ENV, env, 1);
		exec_tmfsrmount(argv);
		perror("tmfs: failed to exec tmfsrmount");
		_exit(1);
	}

	close(fds[0]);
	rv = receive_fd(fds[1]);

	if (!mo->auto_unmount) {
		/* with auto_unmount option tmfsrmount will not exit until 
		   this socket is closed */
		close(fds[1]);
		waitpid(pid, NULL, 0); /* bury zombie */
	}

	return rv;
}

int tmfs_mount_compat22(const char *mountpoint, const char *opts)
{
	struct mount_opts mo;
	memset(&mo, 0, sizeof(mo));
	mo.flags = MS_NOSUID | MS_NODEV;

	return tmfs_mount_tmfsrmount(mountpoint, &mo, opts, 0);
}

static int tmfs_mount_sys(const char *mnt, struct mount_opts *mo,
			  const char *mnt_opts)
{
	char tmp[128];
	const char *devname = "/dev/tmfs";
	char *source = NULL;
	char *type = NULL;
	struct stat stbuf;
	int fd;
	int res;

	if (!mnt) {
		fprintf(stderr, "tmfs: missing mountpoint parameter\n");
		return -1;
	}

	res = stat(mnt, &stbuf);
	if (res == -1) {
		fprintf(stderr ,"tmfs: failed to access mountpoint %s: %s\n",
			mnt, strerror(errno));
		return -1;
	}

	if (!mo->nonempty) {
		res = tmfs_mnt_check_empty("tmfs", mnt, stbuf.st_mode,
					   stbuf.st_size);
		if (res == -1)
			return -1;
	}

	if (mo->auto_unmount) {
		/* Tell the caller to fallback to tmfsrmount because
		   auto-unmount does not work otherwise. */
		return -2;
	}

	fd = open(devname, O_RDWR);
	if (fd == -1) {
		if (errno == ENODEV || errno == ENOENT)
			fprintf(stderr, "tmfs: device not found, try 'modprobe tmfs' first\n");
		else
			fprintf(stderr, "tmfs: failed to open %s: %s\n",
				devname, strerror(errno));
		return -1;
	}

	snprintf(tmp, sizeof(tmp),  "fd=%i,rootmode=%o,user_id=%u,group_id=%u",
		 fd, stbuf.st_mode & S_IFMT, getuid(), getgid());

	res = tmfs_opt_add_opt(&mo->kernel_opts, tmp);
	if (res == -1)
		goto out_close;

	source = malloc((mo->fsname ? strlen(mo->fsname) : 0) +
			(mo->subtype ? strlen(mo->subtype) : 0) +
			strlen(devname) + 32);

	type = malloc((mo->subtype ? strlen(mo->subtype) : 0) + 32);
	if (!type || !source) {
		fprintf(stderr, "tmfs: failed to allocate memory\n");
		goto out_close;
	}

	strcpy(type, mo->blkdev ? "tmfsblk" : "tmfs");
	if (mo->subtype) {
		strcat(type, ".");
		strcat(type, mo->subtype);
	}
	strcpy(source,
	       mo->fsname ? mo->fsname : (mo->subtype ? mo->subtype : devname));

	res = mount(source, mnt, type, mo->flags, mo->kernel_opts);
	if (res == -1 && errno == ENODEV && mo->subtype) {
		/* Probably missing subtype support */
		strcpy(type, mo->blkdev ? "tmfsblk" : "tmfs");
		if (mo->fsname) {
			if (!mo->blkdev)
				sprintf(source, "%s#%s", mo->subtype,
					mo->fsname);
		} else {
			strcpy(source, type);
		}
		res = mount(source, mnt, type, mo->flags, mo->kernel_opts);
	}
	if (res == -1) {
		/*
		 * Maybe kernel doesn't support unprivileged mounts, in this
		 * case try falling back to tmfsrmount
		 */
		if (errno == EPERM) {
			res = -2;
		} else {
			int errno_save = errno;
			if (mo->blkdev && errno == ENODEV &&
			    !tmfs_mnt_check_tmfsblk())
				fprintf(stderr,
					"tmfs: 'tmfsblk' support missing\n");
			else
				fprintf(stderr, "tmfs: mount failed: %s\n",
					strerror(errno_save));
		}

		goto out_close;
	}

#ifndef __NetBSD__
#ifndef IGNORE_MTAB
	if (geteuid() == 0) {
		char *newmnt = tmfs_mnt_resolve_path("tmfs", mnt);
		res = -1;
		if (!newmnt)
			goto out_umount;

		res = tmfs_mnt_add_mount("tmfs", source, newmnt, type,
					 mnt_opts);
		free(newmnt);
		if (res == -1)
			goto out_umount;
	}
#endif /* IGNORE_MTAB */
#endif /* __NetBSD__ */
	free(type);
	free(source);

	return fd;

out_umount:
	umount2(mnt, 2); /* lazy umount */
out_close:
	free(type);
	free(source);
	close(fd);
	return res;
}

static int get_mnt_flag_opts(char **mnt_optsp, int flags)
{
	int i;

	if (!(flags & MS_RDONLY) && tmfs_opt_add_opt(mnt_optsp, "rw") == -1)
		return -1;

	for (i = 0; mount_flags[i].opt != NULL; i++) {
		if (mount_flags[i].on && (flags & mount_flags[i].flag) &&
		    tmfs_opt_add_opt(mnt_optsp, mount_flags[i].opt) == -1)
			return -1;
	}
	return 0;
}

int tmfs_kern_mount(const char *mountpoint, struct tmfs_args *args)
{
	struct mount_opts mo;
	int res = -1;
	char *mnt_opts = NULL;

	memset(&mo, 0, sizeof(mo));
	mo.flags = MS_NOSUID | MS_NODEV;

	if (args &&
	    tmfs_opt_parse(args, &mo, tmfs_mount_opts, tmfs_mount_opt_proc) == -1)
		return -1;

	if (mo.allow_other && mo.allow_root) {
		fprintf(stderr, "tmfs: 'allow_other' and 'allow_root' options are mutually exclusive\n");
		goto out;
	}
	res = 0;
	if (mo.ishelp)
		goto out;

	res = -1;
	if (get_mnt_flag_opts(&mnt_opts, mo.flags) == -1)
		goto out;
	if (mo.kernel_opts && tmfs_opt_add_opt(&mnt_opts, mo.kernel_opts) == -1)
		goto out;
	if (mo.mtab_opts &&  tmfs_opt_add_opt(&mnt_opts, mo.mtab_opts) == -1)
		goto out;

	res = tmfs_mount_sys(mountpoint, &mo, mnt_opts);
	if (res == -2) {
		if (mo.tmfsrmount_opts &&
		    tmfs_opt_add_opt(&mnt_opts, mo.tmfsrmount_opts) == -1)
			goto out;

		if (mo.subtype) {
			char *tmp_opts = NULL;

			res = -1;
			if (tmfs_opt_add_opt(&tmp_opts, mnt_opts) == -1 ||
			    tmfs_opt_add_opt(&tmp_opts, mo.subtype_opt) == -1) {
				free(tmp_opts);
				goto out;
			}

			res = tmfs_mount_tmfsrmount(mountpoint, &mo, tmp_opts, 1);
			free(tmp_opts);
			if (res == -1)
				res = tmfs_mount_tmfsrmount(mountpoint, &mo,
							    mnt_opts, 0);
		} else {
			res = tmfs_mount_tmfsrmount(mountpoint, &mo, mnt_opts, 0);
		}
	}
out:
	free(mnt_opts);
	free(mo.fsname);
	free(mo.subtype);
	free(mo.tmfsrmount_opts);
	free(mo.subtype_opt);
	free(mo.kernel_opts);
	free(mo.mtab_opts);
	return res;
}

TMFS_SYMVER(".symver tmfs_mount_compat22,tmfs_mount@TMFS_2.2");
TMFS_SYMVER(".symver tmfs_unmount_compat22,tmfs_unmount@TMFS_2.2");
